text/0000-read-buf.md

```diff
@@ -0,0 +1,655 @@
+            buf.assume_init(nread);
+            // And indicate that we've written the bytes as well. Unlike `assume_initialized`, this method is safe,
+            // and asserts that the written portion of the buffer does not advance beyond the initialized portion of
+            // the buffer. If we didn't call `assume_init` above, this call could panic.
+            buf.add_filled(nread);
+
+            Ok(())
+        }
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+/// A wrapper around a byte buffer that is incrementally filled and initialized.
+///
+/// This type is a sort of "double cursor". It tracks three regions in the buffer: a region at the beginning of the
+/// buffer that has been logically filled with data, a region that has been initialized at some point but not yet
+/// logically filled, and a region at the end that is fully uninitialized. The filled region is guaranteed to be a
+/// subset of the initialized region.
+///
+/// In summary, the contents of the buffer can be visualized as:
+/// ```not_rust
+/// [             capacity              ]
+/// [ filled |         unfilled         ]
+/// [    initialized    | uninitialized ]
+/// ```
+pub struct ReadBuf<'a> {
+    buf: &'a mut [MaybeUninit<u8>],
+    filled: usize,
+    initialized: usize,
+}
+
+impl<'a> ReadBuf<'a> {
+    /// Creates a new `ReadBuf` from a fully initialized buffer.
+    #[inline]
+    pub fn new(buf: &'a mut [u8]) -> ReadBuf<'a> { ... }
+
+    /// Creates a new `ReadBuf` from a fully uninitialized buffer.
+    ///
+    /// Use `assume_init` if part of the buffer is known to be already inintialized.
+    #[inline]
+    pub fn uninit(buf: &'a mut [MaybeUninit<u8>]) -> ReadBuf<'a> { ... }
+
+    /// Returns the total capacity of the buffer.
+    #[inline]
+    pub fn capacity(&self) -> usize { ... }
+
+    /// Returns a shared reference to the filled portion of the buffer.
+    #[inline]
+    pub fn filled(&self) -> &[u8] { ... }
+
+    /// Returns a mutable reference to the filled portion of the buffer.
+    #[inline]
+    pub fn filled_mut(&mut self) -> &mut [u8] { ... }
+
+    /// Returns a shared reference to the initialized portion of the buffer.
+    ///
+    /// This includes the filled portion.
+    #[inline]
+    pub fn initialized(&self) -> &[u8] { ... }
+
+    /// Returns a mutable reference to the initialized portion of the buffer.
+    ///
+    /// This includes the filled portion.
+    #[inline]
+    pub fn initialized_mut(&mut self) -> &mut [u8] { ... }
+
+    /// Returns a mutable reference to the unfilled part of the buffer without ensuring that it has been fully
+    /// initialized.
+    ///
+    /// # Safety
+    ///
+    /// The caller must not de-initialize portions of the buffer that have already been initialized.
+    #[inline]
+    pub unsafe fn unfilled_mut(&mut self) -> &mut [MaybeUninit<u8>] { ... }
```

Comment:
I would like to propose that the safety section is expanded to the following:

> The caller must not de-initialize portions of the buffer that have already been initialized.
> This includes any bytes in the region marked as uninitialized by `ReadBuf`.

Without a comment like this, it is unclear whether it is sound to put an [`UninitSlice`][1] from the bytes crate into an `ReadBuf` without first initializing the memory.

[1]: https://docs.rs/bytes/0.6/bytes/buf/struct.UninitSlice.html

Reply:
Or simply change that to return an [`Out<[u8]>`](https://docs.rs/uninit/0.4.0/uninit/out_ref/struct.Out.html) **to get rid of the `unsafe` altogether**: even though I like the `ReadBuf` abstraction since it tracks the number of initialized elements (which an `Out<[u8]>` does not), I still think that an `out`-references abstraction is both intuitive and less error-prone than `&mut MaybeUninit<...>`s, and something that is very tied to "writing into potentially-uninit stuff", so _the `&mut MU` part of the `ReadBuf` API ought to be replaced with that_. Being able to, for instance, feature a non-`unsafe` version of this method is just one example of that (and then, if people were to need exactly an `&mut [MU<u8>]`, they could call the `unsafe` `.as_uninit_mut()` method, who has a well-detailed section about how and why writing uninit bytes to the pointee is unsound).